### PR TITLE
Refactor linen Imagenet Example.

### PR DESCRIPTION
**Pull request by mohitreddy1996**
* Refactor ImageNet example implemented using Linen API as part of #231. This PR is similar to #287 and ticks few of the things to do as part of #480 
* Use ml_collections.ConfigDict to configure HPs.
* Add unit test to run a single step of training and evaluation.

### CODE REVIEW

File: linen_examples/imagenet/imagenet_fake_data_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+"""Benchmark for the ImageNet example using fake data for quick perf results."""
+import pathlib
+import time
+
+from absl import flags
+from absl.testing import absltest
+from absl.testing.flagsaver import flagsaver
+import imagenet_main
+from configs import default as default_lib
```
Comment (andsteing) [inline linen_examples/imagenet/imagenet_fake_data_benchmark.py:23]: did you consider adding a testing config?

File: linen_examples/imagenet/imagenet_fake_data_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    flax_root_dir = pathlib.Path(__file__).parents[2]
+    data_dir = str(flax_root_dir) + '/.tfds/metadata'
+
+    start_time = time.time()
+    with tfds.testing.mock_data(num_examples=1024, data_dir=data_dir):
+      imagenet_main.main([])
+    benchmark_time = time.time() - start_time
+
+    self.report_wall_time(benchmark_time)
+    self.report_extras({
```
Comment (andsteing) [inline linen_examples/imagenet/imagenet_fake_data_benchmark.py:66]: What are we testing here? The benchmark setup or the model? If the latter, does the incremental benefit really warrant adding another test file?
(Note that maybe the fake data benchmark was originally added to examples that did not have unit tests for `train_and_evaluate()` in which case they were effectively unit tests.)

**Review by andsteing — CHANGES_REQUESTED** [review]
Thanks for updating this example !!

File: linen_examples/imagenet/README.md [hunk 3]
```diff
@@ -17,8 +17,32 @@ The model should run with other configurations and hardware, but explicitely tes
 
 ### How to run
 
+```shell
+python imagenet_main.py --model_dir=./imagenet
+```
+
+#### Overriding Hyperparameter configurations
+
+Specify a hyperparameter configuration by the means of setting `--config` flag. Configuration flag is defined using
```
[the remark this answers is not in the document]
Comment (mohitreddy1996) [reply linen_examples/imagenet/README.md:26]: Done!

File: linen_examples/imagenet/configs/v100_x8_mixed_precision.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+from configs import default as default_lib
+
+
+def get_config():
+  """Get the hyperparameter configuration to train on 8 x Nvidia V100 GPUs."""
+  # Override default configuration to avoid duplication of field definition.
+  config = default_lib.get_config()
+
+  config.batch_size = 2048
```
[the remark this answers is not in the document]
Comment (mohitreddy1996) [reply linen_examples/imagenet/configs/v100_x8_mixed_precision.py:25]: This was the configuration specified in the README before I made it a config.

File: linen_examples/imagenet/README.md [hunk 5]
```diff
@@ -17,8 +17,32 @@ The model should run with other configurations and hardware, but explicitely tes
...
+
+```shell
+python imagenet_main.py --model_dir=./imagenet --config.num_epochs=100
+```
+
 #### 8 x Nvidia V100 (16GB)
-`python train.py --batch_size=512 --cache=True --model_dir=./imagenet_fp32_bs512`
+```shell
+python imagenet_main.py \
+--model_dir=./imagenet_fp32_bs512 --config=configs/v100_x8.py
```
[the remark this answers is not in the document]
Comment (mohitreddy1996) [reply linen_examples/imagenet/README.md:38]: Done.

File: linen_examples/imagenet/imagenet_benchmark.py [hunk 6]
```diff
@@ -40,13 +41,16 @@ class ImagenetBenchmark(Benchmark):
   def test_8x_v100_half_precision(self):
     """Run ImageNet on 8x V100 GPUs in half precision for 2 epochs."""
     model_dir = self.get_tmp_model_dir()
-    FLAGS.batch_size = 2048
-    FLAGS.half_precision = True
-    FLAGS.num_epochs = 2
+    config = default_lib.get_config()
```
[the remark this answers is not in the document]
Comment (mohitreddy1996) [reply linen_examples/imagenet/imagenet_benchmark.py:44]: Done.

File: linen_examples/imagenet/resnet_v1.py [hunk 7]
```diff
@@ -12,8 +12,7 @@
 # See the License for the specific language governing permissions and
```
[the remark this answers is not in the document]
Comment (mohitreddy1996) [reply linen_examples/imagenet/resnet_v1.py:12]: I had renamed the file as resnet_v1.py in Flax/examples as well and also it is consistent with the internal version.

File: linen_examples/imagenet/resnet_v1_test.py [hunk 8]
```diff
@@ -12,40 +12,33 @@
...
-    self.assertEqual(y.shape, (8, 1000))
+  def test_resnet_v1_model(self):
+    """Tests ResNet V1 model definition."""
+    rng = jax.random.PRNGKey(0)
+    model_def = models.ResNet50(num_classes=10, dtype=jnp.float32)
+    variables = model_def.init(
+        rng, jnp.ones((8, 224, 224, 3), jnp.float32))
 
+    self.assertLen(variables, 2)
+    self.assertLen(variables['params'], 19)
```
[the remark this answers is not in the document]
Comment (mohitreddy1996) [reply linen_examples/imagenet/resnet_v1_test.py:41]: Done.

File: linen_examples/imagenet/resnet_v1_test.py [hunk 9]
```diff
@@ -12,40 +12,33 @@
...
-    super().setUp()
-    # Make sure tf does not allocate gpu memory.
-    tf.config.experimental.set_visible_devices([], 'GPU')
+class ResNetV1Test(absltest.TestCase):
+  """Test cases for ResNet v1 model."""
 
-  def test_create_model(self):
-    variables = train.initialized(random.PRNGKey(0), 224)
-    x = random.normal(random.PRNGKey(1), (8, 224, 224, 3))
-    y = train.model(train=False).apply(variables, x)
```
[the remark this answers is not in the document]
Comment (mohitreddy1996) [reply linen_examples/imagenet/resnet_v1_test.py:46]: Moved it to the _lib_test.py

File: linen_examples/imagenet/imagenet_lib.py [hunk 10]
```diff
@@ -38,51 +38,22 @@
...
 
-flags.DEFINE_bool(
-    'half_precision', default=False,
-    help=('If bfloat16/float16 should be used instead of float32.'))
+# enable jax omnistaging
+jax.config.enable_omnistaging()
 
 
-def model(**kwargs):
+def model(half_precision, **kwargs):
```
[the remark this answers is not in the document]
Comment (mohitreddy1996) [reply linen_examples/imagenet/imagenet_lib.py:53]: Good suggestion, done!

File: linen_examples/imagenet/imagenet_lib.py [hunk 11]
```diff
@@ -238,30 +213,29 @@ def sync_batch_stats(state):
   return state.replace(batch_stats=avg(state.batch_stats))
 
 
-def main(argv):
-  if len(argv) > 1:
-    raise app.UsageError('Too many command-line arguments.')
+def train_and_evaluate(model_dir: str, config: ml_collections.ConfigDict):
```
[the remark this answers is not in the document]
Comment (mohitreddy1996) [reply linen_examples/imagenet/imagenet_lib.py:216]: Thanks!

File: linen_examples/imagenet/imagenet_fake_data_benchmark.py (hunk 2 above)
Comment (mohitreddy1996) [reply linen_examples/imagenet/imagenet_fake_data_benchmark.py:66]: This is a benchmark used by an external team to benchmark JAX and Flax compilation so a minimal setup is preferred.

File: linen_examples/imagenet/imagenet_fake_data_benchmark.py (hunk 1 above)
Comment (mohitreddy1996) [reply linen_examples/imagenet/imagenet_fake_data_benchmark.py]: Good idea, thanks!

File: linen_examples/imagenet/resnet_v1.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (andsteing) [reply linen_examples/imagenet/resnet_v1.py]: I would vote for keeping `models.py` for inter-example consistency. Also updated #231 that now explicitly mentions `models.py` as an additional file.

File: linen_examples/imagenet/imagenet_fake_data_benchmark.py (hunk 2 above)
Comment (andsteing) [reply linen_examples/imagenet/imagenet_fake_data_benchmark.py:56]: Ah, so it's about having the benchmark interface for automated testing.

**Review by andsteing — APPROVED** [review]

File: linen_examples/imagenet/resnet_v1.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mohitreddy1996) [reply linen_examples/imagenet/resnet_v1.py]: Thanks for pointing this out!

**Pull request merged**